# Basic SQL Queries: Flights Database

## The Database can be found [here](https://drive.google.com/file/d/1uGq4BEst0QUzvX3mbTOYQeyYu87wEOCW/view?usp=drive_link)

Below is an Entity Relationship Diagram (ERD) of the flights database:

<img src="https://github.com/Explore-AI/Pictures/blob/master/flights_ER_diagram.png?raw=true" width=40%/>

As can be seen from the ERD, the flights database consists of four tables:
- **flights**: all domestic flights in the USA in 2008
- **carriers**: lookup table for all the carriers
- **planes**: lookup table for all the planes 
- **airports**: lookup table for all the airports

For this challenge, I will apply the use of Basic SQL queries to gain insight into the flights database. 

## Loading the database

Before producing basic SQL queries, we prepare the SQL environment by loading in the magic command `%load_ext sql`

In [ ]:
%load_ext sql

Next we load in the database.

In [5]:
%%sql 

sqlite:///flights.db

In [ ]:
%%time %%sql

Exploring the database 

Let's go ahead and take a look at the composition of each table in the flights database

In [7]:
%%sql 

PRAGMA table_info(flights);

 * sqlite:///flights.db
Done.


[(0, 'index', 'INTEGER', 0, 'NULL', 0),
 (1, 'Date', 'longtext', 0, None, 0),
 (2, 'DayOfWeek', 'INTEGER', 0, 'NULL', 0),
 (3, 'DepTime', 'double', 0, 'NULL', 0),
 (4, 'CRSDepTime', 'longtext', 0, None, 0),
 (5, 'ArrTime', 'double', 0, 'NULL', 0),
 (6, 'CRSArrTime', 'longtext', 0, None, 0),
 (7, 'UniqueCarrier', 'longtext', 0, None, 0),
 (8, 'FlightNum', 'longtext', 0, None, 0),
 (9, 'TailNum', 'longtext', 0, None, 0),
 (10, 'ActualElapsedTime', 'double', 0, 'NULL', 0),
 (11, 'CRSElapsedTime', 'longtext', 0, None, 0),
 (12, 'AirTime', 'longtext', 0, None, 0),
 (13, 'ArrDelay', 'double', 0, 'NULL', 0),
 (14, 'DepDelay', 'double', 0, 'NULL', 0),
 (15, 'Origin', 'longtext', 0, None, 0),
 (16, 'Dest', 'longtext', 0, None, 0),
 (17, 'Distance', 'INTEGER', 0, 'NULL', 0),
 (18, 'TaxiIn', 'longtext', 0, None, 0),
 (19, 'TaxiOut', 'longtext', 0, None, 0),
 (20, 'Cancelled', 'longtext', 0, None, 0),
 (21, 'CancellationCode', 'longtext', 0, None, 0),
 (22, 'Diverted', 'longtext', 0, None, 0),
 (23, 'CarrierDelay', 'longtext', 0, None, 0),
 (24, 'WeatherDelay', 'longtext', 0, None, 0),
 (25, 'NASDelay', 'longtext', 0, None, 0),
 (26, 'SecurityDelay', 'longtext', 0, None, 0),
 (27, 'LateAircraftDelay', 'longtext', 0, None, 0)]

In [5]:
%%sql 

PRAGMA table_info(carriers);

 * sqlite:///flights.db
Done.


[(0, 'Code', 'varchar(10)', 0, 'NULL', 0),
 (1, 'Description', 'varchar(100)', 0, 'NULL', 0)]

In [6]:
%%sql 

PRAGMA table_info(airports);

 * sqlite:///flights.db
Done.


[(0, 'iata', 'varchar(10)', 0, 'NULL', 0),
 (1, 'airport', 'varchar(100)', 0, 'NULL', 0),
 (2, 'city', 'varchar(100)', 0, 'NULL', 0),
 (3, 'state', 'varchar(50)', 0, 'NULL', 0),
 (4, 'country', 'varchar(50)', 0, 'NULL', 0),
 (5, 'lat', 'float(24,0)', 0, 'NULL', 0),
 (6, 'long', 'float(24,0)', 0, 'NULL', 0)]

In [7]:
%%sql 

PRAGMA table_info(planes);

 * sqlite:///flights.db
Done.


[(0, 'tailnum', 'varchar(10)', 0, 'NULL', 0),
 (1, 'type', 'varchar(100)', 0, 'NULL', 0),
 (2, 'manufacturer', 'varchar(100)', 0, 'NULL', 0),
 (3, 'issue_date', 'varchar(100)', 0, 'NULL', 0),
 (4, 'model', 'varchar(50)', 0, 'NULL', 0),
 (5, 'status', 'varchar(100)', 0, 'NULL', 0),
 (6, 'aircraft_type', 'varchar(100)', 0, 'NULL', 0),
 (7, 'engine_type', 'varchar(100)', 0, 'NULL', 0),
 (8, 'year', 'longtext', 0, None, 0)]

## Questions on Basic SQL Queries 

**Question 1**

How many different carriers are there in total in the database?

In [8]:
%%sql
SELECT * FROM carriers
LIMIT 15;

 * sqlite:///flights.db
Done.


[('02Q', 'Titan Airways\r'),
 ('04Q', 'Tradewind Aviation\r'),
 ('05Q', 'Comlux Aviation, AG\r'),
 ('06Q', 'Master Top Linhas Aereas Ltd.\r'),
 ('07Q', 'Flair Airlines Ltd.\r'),
 ('09Q', 'Swift Air, LLC\r'),
 ('0BQ', 'DCA\r'),
 ('0CQ', 'ACM AIR CHARTER GmbH\r'),
 ('0FQ', 'Maine Aviation Aircraft Charter, LLC\r'),
 ('0GQ', 'Inter Island Airways, d/b/a Inter Island Air\r'),
 ('0HQ', 'Polar Airlines de Mexico d/b/a Nova Air\r'),
 ('0J', 'JetClub AG\r'),
 ('0JQ', 'Vision Airlines\r'),
 ('0KQ', 'Mokulele Flight Services, Inc.\r'),
 ('0LQ', 'Metropix UK, LLP.\r')]

In [9]:
%%sql
SELECT COUNT(Code AND Description) FROM carriers;

 * sqlite:///flights.db
Done.


[(1491,)]

### ANSWER: 1491

**Question 2** 

How long was the longest delay before departure?

In [10]:
%%sql
SELECT MAX(DepDelay) FROM flights

 * sqlite:///flights.db
Done.


[(1355.0,)]

### ANSWER: 1355 min

**Question 3** 

How many flights departed on the 28th of January 2008?

In [11]:
%%sql
SELECT COUNT(*) FROM flights WHERE Date = '2008/1/28'

 * sqlite:///flights.db
Done.


[(20149,)]

### ANSWER: 20149

**Question 4**
 
 What is the distance between Midway Airport (MDW) and Houston Airport (HOU)?

In [12]:
%%sql
SELECT Origin, Dest, Distance FROM flights WHERE Origin = 'MDW' AND Dest = 'HOU'
LIMIT 1

 * sqlite:///flights.db
Done.


[('MDW', 'HOU', 937)]

### ANSWER: 937km

**Question 5**

Which day of the week had the highest number of cancelled flights?(Note 1 = cancelled , 0 = not cancelled;)

In [18]:
%%sql
SELECT DayOfWeek, COUNT(DayOfWeek) FROM flights WHERE Cancelled = 1  GROUP BY DayOfWeek

 * sqlite:///flights.db
Done.


[(1, 2617), (2, 2993), (3, 2645), (4, 3093), (5, 2049), (6, 1984), (7, 1927)]

### ANSWER: Thursday(4)

**Question 6**

How many airports have the word "International" in their name?

In [14]:
%%sql
SELECT COUNT(airport) FROM airports WHERE airport LIKE '%International%'

 * sqlite:///flights.db
Done.


[(124,)]

### ANSWER:124

**Question 7**

What is the most produced model for the manufacturer "BOEING"?

In [15]:
%%sql
SELECT model, SUM (DISTINCT model) FROM planes WHERE manufacturer = "BOEING" GROUP BY model

 * sqlite:///flights.db
Done.


[('717-200', 717.0),
 ('737-230', 737.0),
 ('737-236', 737.0),
 ('737-282', 737.0),
 ('737-282C', 737.0),
 ('737-2P6', 737.0),
 ('737-2X6C', 737.0),
 ('737-2Y5', 737.0),
 ('737-301', 737.0),
 ('737-317', 737.0),
 ('737-322', 737.0),
 ('737-33A', 737.0),
 ('737-3A4', 737.0),
 ('737-3B7', 737.0),
 ('737-3G7', 737.0),
 ('737-3H4', 737.0),
 ('737-3K2', 737.0),
 ('737-3L9', 737.0),
 ('737-3Q8', 737.0),
 ('737-3S3', 737.0),
 ('737-3T5', 737.0),
 ('737-3TO', 737.0),
 ('737-3Y0', 737.0),
 ('737-401', 737.0),
 ('737-490', 737.0),
 ('737-4B7', 737.0),
 ('737-4Q8', 737.0),
 ('737-4S3', 737.0),
 ('737-522', 737.0),
 ('737-524', 737.0),
 ('737-5H4', 737.0),
 ('737-705', 737.0),
 ('737-724', 737.0),
 ('737-73A', 737.0),
 ('737-76N', 737.0),
 ('737-76Q', 737.0),
 ('737-790', 737.0),
 ('737-7AD', 737.0),
 ('737-7BD', 737.0),
 ('737-7H4', 737.0),
 ('737-7Q8', 737.0),
 ('737-824', 737.0),
 ('737-832', 737.0),
 ('737-890', 737.0),
 ('737-8FH', 737.0),
 ('737-924', 737.0),
 ('737-924ER', 737.0),
 ('737-990', 737.0),
 ('747-2B5F', 747.0),
 ('747-422', 747.0),
 ('747-451', 747.0),
 ('757-212', 757.0),
 ('757-222', 757.0),
 ('757-223', 757.0),
 ('757-224', 757.0),
 ('757-225', 757.0),
 ('757-231', 757.0),
 ('757-232', 757.0),
 ('757-23N', 757.0),
 ('757-251', 757.0),
 ('757-26D', 757.0),
 ('757-2B7', 757.0),
 ('757-2G7', 757.0),
 ('757-2Q8', 757.0),
 ('757-2S7', 757.0),
 ('757-324', 757.0),
 ('757-33N', 757.0),
 ('757-351', 757.0),
 ('767-201', 767.0),
 ('767-223', 767.0),
 ('767-224', 767.0),
 ('767-2B7', 767.0),
 ('767-322', 767.0),
 ('767-323', 767.0),
 ('767-324', 767.0),
 ('767-332', 767.0),
 ('767-33A', 767.0),
 ('767-3CB', 767.0),
 ('767-3G5', 767.0),
 ('767-3P6', 767.0),
 ('767-424ER', 767.0),
 ('767-432ER', 767.0),
 ('777-222', 777.0),
 ('777-224', 777.0),
 ('777-232', 777.0),
 ('777-232LR', 777.0)]

### ANSWER: 757-222

**Question 8**

How many planes landed at Los Angeles International Airport?

In [16]:
%%sql
SELECT airport, COUNT(Dest) FROM airports
LEFT JOIN flights ON airports.iata = flights.Dest
WHERE airport = 'Los Angeles International'

 * sqlite:///flights.db
Done.


[('Los Angeles International', 18964)]

### ANSWER: 18964